src/client/texturesource.cpp

```diff
@@ -1848,58 +1856,105 @@ bool TextureSource::generateImagePart(std::string_view part_of_name,
+	auto a_new_255{src_a * 255 + (255 - src_a) * dst_a};
+	dst.r = (dst.r * (255 - src_a) * dst_a + src.r * src_a * 255) / a_new_255;
+	dst.g = (dst.g * (255 - src_a) * dst_a + src.g * src_a * 255) / a_new_255;
+	dst.b = (dst.b * (255 - src_a) * dst_a + src.b * src_a * 255) / a_new_255;
+	dst_a = a_new_255 / 255;
+	dst_col.set(dst_a, dst.r, dst.g, dst.b);
 }
 
-/*
-	Draw an image on top of another one, using the alpha channel of the
-	source image
+}  // namespace
 
-	This exists because IImage::copyToWithAlpha() doesn't seem to always
-	work.
-*/
 template<bool overlay>
-static void blit_with_alpha(video::IImage *src, video::IImage *dst,
-		v2s32 src_pos, v2s32 dst_pos, v2u32 size)
+void blit_with_alpha(video::IImage *src, video::IImage *dst, v2s32 dst_pos,
+	v2u32 size)
 {
-	auto src_dim = src->getDimension();
-	auto dst_dim = dst->getDimension();
-
+	if (dst->getColorFormat() != video::ECF_A8R8G8B8)
+		throw BaseException("blit_with_alpha() supports only ECF_A8R8G8B8 "
+			"destination images.");
+
+	auto src_dim{src->getDimension()};
+	auto dst_dim{dst->getDimension()};
+	bool drop_src{false};
+	if (src->getColorFormat() != video::ECF_A8R8G8B8) {
+		video::IVideoDriver *driver{RenderingEngine::get_video_driver()};
+		video::IImage *src_converted{driver->createImage(video::ECF_A8R8G8B8,
+			src_dim)};
+		if (!src_converted)
+			throw BaseException("blit_with_alpha() failed to convert the "
+				"source image to ECF_A8R8G8B8.");
+		src->copyTo(src_converted);
+		src = src_converted;
+		drop_src = true;
+	}
+	video::SColor *pixels_src
+		{reinterpret_cast<video::SColor *>(src->getData())};
+	video::SColor *pixels_dst
+		{reinterpret_cast<video::SColor *>(dst->getData())};
 	// Limit y and x to the overlapping ranges
 	// s.t. the positions are all in bounds after offsetting.
-	for (u32 y0 = std::max(0, -dst_pos.Y);
-			y0 < std::min<s64>({size.Y, src_dim.Height, dst_dim.Height - (s64) dst_pos.Y});
-			++y0)
-	for (u32 x0 = std::max(0, -dst_pos.X);
-			x0 < std::min<s64>({size.X, src_dim.Width, dst_dim.Width - (s64) dst_pos.X});
-			++x0)
-	{
-		s32 src_x = src_pos.X + x0;
-		s32 src_y = src_pos.Y + y0;
-		s32 dst_x = dst_pos.X + x0;
-		s32 dst_y = dst_pos.Y + y0;
-		video::SColor src_c = src->getPixel(src_x, src_y);
-		video::SColor dst_c = dst->getPixel(dst_x, dst_y);
-		if (!overlay || (dst_c.getAlpha() == 255 && src_c.getAlpha() != 0)) {
-			dst_c = blitPixel(src_c, dst_c, src_c.getAlpha());
-			dst->setPixel(dst_x, dst_y, dst_c);
+	u32 x_start{static_cast<u32>(std::max(0, -dst_pos.X))};
+	u32 y_start{static_cast<u32>(std::max(0, -dst_pos.Y))};
+	u32 x_end{static_cast<u32>(std::min<s64>({size.X, src_dim.Width,
+		dst_dim.Width - (s64) dst_pos.X}))};
+	u32 y_end{static_cast<u32>(std::min<s64>({size.Y, src_dim.Height,
+		dst_dim.Height - (s64) dst_pos.Y}))};
+	for (u32 y0{y_start}; y0 < y_end; ++y0) {
+		size_t i_src{y0 * src_dim.Width + x_start};
+		size_t i_dst{(dst_pos.Y + y0) * dst_dim.Width + dst_pos.X + x_start};
```

Comment:
Here too. This is unreadable and cannot prevent from over- and overflows (which is what would be more important here).